### PR TITLE
rose macro: add reporter macros

### LINKED ISSUE #1687: rose macro: new type (report)
At present there are two types of rose macros, `Transform` and `Validate`.
I've been working on 'printing' information from config files in /human readable/ form.

I don't want these macros run when a user tries to `validate_all` but they're not strictly Transform macros either as they don't alter the config.

Would it be possible to have a third type of macro, similar to Validate but not included in validate all such as a `Report` type of similar ?

### PR DESCRIPTION
**Pull request by oliver-sanders**
Closes #1687 

Adds a new type of macro for reporting on application configurations, the main purpose of this is so that these macros don't get run by `rose macro -V`.

@matthewrmshin @arjclark Please Review

### CODE REVIEW

File: lib/python/rose/macro.py [hunk 1]
```diff
@@ -690,16 +691,19 @@ def check_config_integrity(app_config):
...
+def report_config(app_config, meta_config, run_macro_list, modules,
+                  macro_info_tuples, opt_non_interactive=False,
+                  optional_config_name=None, optional_values=None,
+                  validate_mode=True):
+    """Run report/validator custom macros on the config and return problems
+    (in the case of validator macros)."""
     if optional_values is None:
         optional_values = {}
     macro_problem_map = {}
+    macro_method = VALIDATE_METHOD if validate_mode else REPORT_METHOD
```
Comment (matthewrmshin) [inline lib/python/rose/macro.py]: Please consider using a normal `if...else` block here.

File: doc/rose-rug-advanced-tutorials-macro.html [hunk 2]
```diff
@@ -589,6 +589,21 @@ <h3 class="alwayshidden">Example Metadata Option</h3>
...
             <div class="slide">
+              <h2 id="reporter-macros">Reporter Macros</h2>
+              <p>There are also reporter macros which can be used
+              where you need to output some information about a
+              configuration. A reporter macro takes the same form
+              as validator and transform macros but does not require
+              a return value.</p>
+
+              <pre class="prettyprint">
+    def report(self, config, meta_config=None):
```
Comment (arjclark) [inline doc/rose-rug-advanced-tutorials-macro.html]: Please make the example here part of the tutorial itself so someone can work through this against the rest of the tutorial. At present this is just a copy of what's in the API doc and not relevant to the tutorial itself - there's no snowflakes here. I recommend something like "planet.PlanetReporter" to write the planet name to a file.

Comment (arjclark) [inline doc/rose-rug-advanced-tutorials-macro.html]: Once that change is made I'll work through the tutorial and make sure everything's fine with Rose edit integration etc as that isn't checked in the test battery.

**Comment by oliver-sanders** [comment]
@arjclark Have now added a (short) tutorial.

File: doc/etc/rose-rug-advanced-tutorials-macro/planet.py.report.html [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            print 'Please choose a planet other than Earth.'
+            return
+        constellation = self.get_planet_info(planet)
+        if not constellation:
+            print 'Could not find horoscope entry for {0}'.format(planet)
+            return
+        else:
+            print (
+                '{planet} is currently passing through {constellation}.\n'
+                'You should {tosh} today.'
```
Comment (arjclark) [inline doc/etc/rose-rug-advanced-tutorials-macro/planet.py.report.html]: Can you pick something other than "tosh" e.g. "statement" or "advice". If nothing else it's colloquial...

Comment (oliver-sanders) [inline doc/etc/rose-rug-advanced-tutorials-macro/planet.py.report.html]: Done.

**Pull request merged**